torch/nn/parallel/distributed.py

```diff
@@ -1554,7 +1554,18 @@ def _check_global_requires_backward_grad_sync(self, is_joined_rank):
         work = dist.all_reduce(
             requires_sync_tensor, group=self.process_group, async_op=True
         )
-        return work
+
+        # (kwen2501) This if condition is a plain translation of previous
+        # behavior, i.e. in the `is_joined_rank=False` case, `work.wait()`
+        # is not called and it doesn't care about the result. I am guessing
+        # that it just wants to fire a matching all-reduce and does not want
+        # the main stream to wait.
+        if is_joined_rank:
+            work.wait()
+            should_sync_backwards = requires_sync_tensor.item() != 0
+            return should_sync_backwards
+        else:
+            return None  # Return value is not/should not be used.
```

Comment:
would this cause problem when the flag `is_joined_rank` is not same across all ranks and DDP is also using sub-PG scenario? Because, even if we create zero tensors in L1552. We might still get a non-zero results after all-reduce, no?

Reply:
Right. That's why when `is_joined_rank=True` and we created a zero tensor, we would have to do work.wait() to get the all-reduce result.

Reply:
In the case `is_joined_rank=False`, we don't care about the all-reduce result, see [here](https://github.com/pytorch/pytorch/blob/94f19b1fe6ebc6001edfe9948d894b29c9f41f35/torch/nn/parallel/distributed.py#L1424C81-L1424C81)

Reply:
I guess my question is that, would this cause hang? Are all ranks calling all-reduce have the same value of `is_joined_rank`?

Reply:
They don't need to have the same value for `is_joined_rank`. The purpose is for all ranks to call all-reduce.